libraries/Matter/examples/MatterOccupancySensor/MatterOccupancySensor.ino

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * This example is an example code that will create a Matter Device which can be
+ * commissioned and controlled from a Matter Environment APP.
+ * Additionally the ESP32 will send debug messages indicating the Matter activity.
+ * Turning DEBUG Level ON may be useful to following Matter Accessory and Controller messages.
+ *
+ * The example will create a Matter Occupancy Sensor Device.
+ * The Occupancy Sensor will be simulated to change its state every 2 minutes.
+ *
+ * The onboard button can be kept pressed for 5 seconds to decommission the Matter Node.
+ * The example will also show the manual commissioning code and QR code to be used in the Matter environment.
+ *
+ */
+
+// Matter Manager
+#include <Matter.h>
+#include <WiFi.h>
+
+// List of Matter Endpoints for this Node
+// Matter Occupancy Sensor Endpoint
+MatterOccupancySensor OccupancySensor;
+
+// set your board USER BUTTON pin here - decommissioning only
+const uint8_t buttonPin = BOOT_PIN;  // Set your pin here. Using BOOT Button.
+
+// WiFi is manually set and started
+const char *ssid = "your-ssid";          // Change this to your WiFi SSID
+const char *password = "your-password";  // Change this to your WiFi password
+
+// Button control
+uint32_t button_time_stamp = 0;                // debouncing control
+bool button_state = false;                     // false = released | true = pressed
+const uint32_t decommissioningTimeout = 5000;  // keep the button pressed for 5s, or longer, to decommission
+
+void setup() {
+  // Initialize the USER BUTTON (Boot button) that will be used to decommission the Matter Node
+  // The button will also be used to manually toggle the Occupancy Sensor state
+  pinMode(buttonPin, INPUT_PULLUP);
+
+  Serial.begin(115200);
+
+  // Manually connect to WiFi
+  WiFi.begin(ssid, password);
+  // Wait for connection
+  while (WiFi.status() != WL_CONNECTED) {
+    delay(500);
+    Serial.print(".");
+  }
+  Serial.println();
+
+  // set initial occupancy sensor state as false and connected to a PIR sensor type (default)
```

Comment:
Sketch suggests that is not actually PIR that is being used (PIR produces short pulses when it detects movement), but rather mmWave type sensor that detects body and pulse is long (as long as the body is being detected in range). I base this on the fact that you simulate 2minutes long pulses

Reply:
I also wonder how would an ultrasonic sensor be implemented, given that it produces a range value and not a boolean pulse.

Reply:
Under Matter perspective, it doesn't matter.
The sketch shall deal with whatever occupancy technology is used.
All Matter cares is about a boolean state :: Occupied or Vacant.

PIR / Ultrasonic is just an information that may or nor be displayed by the APP.

Reply:
> I also wonder how would an ultrasonic sensor be implemented, given that it produces a range value and not a boolean pulse.

I tested it with 10 seconds and a button used to toggle the Matter Boolean Occupancy flag.
Matter doesn't deal with any working Sensor functionality. It just cares about the Boolean Flag.

Reply:
ok